### PR TITLE
Add git-extract admin

### LINKED ISSUE #14537: Create admin page to add entries to git extraction queue
We need an admin page so we can add add-ons to the git extraction queue, so they work in code-manager.

### PR DESCRIPTION
**Pull request by wagnerand**
Fix #14537.

### CODE REVIEW

**Comment by wagnerand** [comment]
@willdurand might want to take a look at this as well.

File: src/olympia/addons/tests/test_admin.py [hunk 1]
```diff
@@ -528,6 +532,63 @@ def test_version_pagination(self):
...
+        addon1 = addon_factory()
+        addon2 = addon_factory()
+
+        addons = Addon.objects.filter(
+            pk__in=(addon1.pk, addon2.pk))
+        addon_admin = AddonAdmin(Addon, admin.site)
+        request = RequestFactory().get('/')
+        request.user = user_factory()
+        core.set_user(request.user)
+        request._messages = default_messages_storage(request)
```
Comment (eviljeff) [inline src/olympia/addons/tests/test_admin.py:545]: so what's this for?  we're not checking activity log or the message output.  (should we be?  `message_user` is being called)

**Review by eviljeff — APPROVED** [review]

File: src/olympia/addons/admin.py [hunk 2]
```diff
@@ -248,6 +270,34 @@ def save_model(self, request, obj, form, change):
             log.info('Addon "%s" status changed to: %s' % (
                 obj.slug, form.cleaned_data['status']))
 
+    def git_extract_action(self, request, qs):
+        addons = []
```
[the remark this answers is not in the document]
Comment (wagnerand) [reply src/olympia/addons/admin.py:274]: I took this from https://github.com/mozilla/addons-server/blob/master/src/olympia/users/admin.py#L252

File: src/olympia/addons/admin.py [hunk 3]
```diff
@@ -248,6 +270,34 @@ def save_model(self, request, obj, form, change):
...
+                              '"%(addons)s".' % kw))
+    git_extract_action.short_description = "Git-Extract"
+
+    def git_extract_view(self, request, object_id, extra_context=None):
+        if request.method != 'POST':
+            return HttpResponseNotAllowed(['POST'])
+
+        obj = self.get_object(request, unquote(object_id))
+        if obj is None:
+            raise Http404()
```
[the remark this answers is not in the document]
Comment (wagnerand) [reply src/olympia/addons/admin.py:290]: Taken from https://github.com/mozilla/addons-server/blob/master/src/olympia/users/admin.py#L179:L179

File: src/olympia/addons/admin.py [hunk 4]
```diff
@@ -248,6 +270,34 @@ def save_model(self, request, obj, form, change):
...
+
+        if not acl.action_allowed(request, amo.permissions.ADDONS_EDIT):
+            return HttpResponseForbidden()
```
[the remark this answers is not in the document]
Comment (wagnerand) [reply src/olympia/addons/admin.py:293]: https://github.com/mozilla/addons-server/blob/master/src/olympia/users/admin.py#L181:L181

File: src/olympia/addons/admin.py [hunk 5]
```diff
@@ -248,6 +270,34 @@ def save_model(self, request, obj, form, change):
             log.info('Addon "%s" status changed to: %s' % (
                 obj.slug, form.cleaned_data['status']))
 
+    def git_extract_action(self, request, qs):
+        addons = []
+        for addon in qs:
+            GitExtractionEntry.objects.create(addon=addon)
+            addons.append(force_text(addon))
+        kw = {'addons': u', '.join(addons)}
```
[the remark this answers is not in the document]
Comment (wagnerand) [reply src/olympia/addons/admin.py:278]: The year where `u'` is still valid :)

File: src/olympia/addons/admin.py [hunk 6]
```diff
@@ -248,6 +270,34 @@ def save_model(self, request, obj, form, change):
...
+
+        obj = self.get_object(request, unquote(object_id))
+        if obj is None:
+            raise Http404()
+
+        if not acl.action_allowed(request, amo.permissions.ADDONS_EDIT):
+            return HttpResponseForbidden()
+
+        self.git_extract_action(
+            request, Addon.objects.filter(pk=obj.pk))
```
[the remark this answers is not in the document]
Comment (wagnerand) [reply src/olympia/addons/admin.py:296]: https://github.com/mozilla/addons-server/blob/master/src/olympia/users/admin.py?rgh-link-date=2020-10-20T15%3A52%3A59Z#L184:L185

File: src/olympia/addons/tests/test_admin.py (hunk 1 above)
Comment (wagnerand) [reply src/olympia/addons/tests/test_admin.py:545]: I took this from https://github.com/mozilla/addons-server/blob/master/src/olympia/users/tests/test_admin.py#L269:L269

File: src/olympia/addons/admin.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (eviljeff) [reply src/olympia/addons/admin.py:296]: in a case where the function being called _does_ need a queryset.  This doesn't - I checked before I said it.

File: src/olympia/addons/admin.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (eviljeff) [reply src/olympia/addons/admin.py:278]: Not a year where python 2 is supported then

File: src/olympia/addons/tests/test_admin.py (hunk 1 above)
Comment (eviljeff) [reply src/olympia/addons/tests/test_admin.py:545]: that test _does_ check the activity log.  (And should check the message output really, if it's not it's been overlooked)

File: src/olympia/addons/admin.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (wagnerand) [reply src/olympia/addons/admin.py]: 0ad536e88d

File: src/olympia/addons/admin.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (wagnerand) [reply src/olympia/addons/admin.py:290]: 0ad536e88d

File: src/olympia/addons/admin.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (wagnerand) [reply src/olympia/addons/admin.py]: 0ad536e88d

File: src/olympia/addons/admin.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (eviljeff) [reply src/olympia/addons/admin.py:290]: meh.  returning forbidden before exposing if an object exists is good opsec.  Doesn't matter much in this case though.

File: src/olympia/addons/admin.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (wagnerand) [reply src/olympia/addons/admin.py]: 0ad536e88d

File: src/olympia/addons/admin.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (wagnerand) [reply src/olympia/addons/admin.py]: 0ad536e88d

File: src/olympia/addons/admin.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (eviljeff) [reply src/olympia/addons/admin.py]: 🤷

File: src/olympia/addons/tests/test_admin.py (hunk 1 above)
Comment (wagnerand) [reply src/olympia/addons/tests/test_admin.py:545]: There is no immediate activity log for this action. Should I remove this line then?

Comment (wagnerand) [reply src/olympia/addons/tests/test_admin.py:545]: The test fails without that line.

Comment (eviljeff) [reply src/olympia/addons/tests/test_admin.py:545]: I would and see if it's needed.  It might fail, but then we know it's doing something important.

**Pull request merged**